Add description in the perfect aspect to the Task model

ivr42 opened the pull request:
- field `description_for_message` was added to the following models: `Task`, `TaskResponse`
 - field `task_description_for_message` was added to the following models `ReportSummaryResponse`, `TaskInfoResponse`, `FullReportDto`
 - alembic migration was created
 - `description_for_message` was added to the `create_task` and `update_task` methods of the `TaskService` class in `core/services/task_service.py`
 - tasks.json: descriptions in perfect aspect were added to all records
 - api/routers/task.py: docstrings was changed

Описание задачи: [Notion](https://www.notion.so/f8689b3d8a02443589375859d446e763)

ivr42 committed:
Add description in the perfect aspect to the Task model

 - field `description_perfect_aspect` was added to the following
   models: `Task`, `TaskResponse`
 - field `task_description_perfect_aspect` was added to the following
   models `ReportSummaryResponse`, `TaskInfoResponse`, `FullReportDto`
 - alembic migration was created
 - `description_perfect_aspect` was added to the `create_task` method
   of the `TaskService` class in `core/services/task_service.py`
 - tasks.json: descriptions in perfect aspect were added to all records
 - api/routers/task.py: docstrings was changed

ADesBiysk commented on src/api/response_models/report.py:
```diff
@@ -44,6 +44,7 @@ class ReportSummaryResponse(BaseModel):
     user_surname: str
     task_id: UUID
     task_description: str
+    task_description_perfect_aspect: str
```
Название смущает. Сомневаюсь, что здесь важен вид глагола. Тем более, заказчики могут изменить предложение и тогда требования к фразе изменятся. Я бы назвал: description_for_message.
Предлагаю обсудить этот момент перед исправлением.

ivr42 replied:
Смотри, задание звучит так: "Добавить описание задания в соответствующем падеже".

> Disclaimer 
> Слово "падеж" тут употреблено не совсем корректно: падежи — это для существительных, а тут глаголы, и то, что они хотят называется "совершенный вид глагола" (пришлось вспоминать уроки русского языка — вот уж никогда бы не подумал, что для программирования мне понадобится русский язык 😀 ). Отсюда и название поля: 'perfect aspect' — это, дословно, совершенный вид глагола на английском.  

Что мы имеем сейчас: у нас есть поле `description`, в котором задание сформулировано в повелительном наклонении: "**Помой** пол" и мы вводим второе поле, в котором задание сформулировано в совершенной форме:  "**помыть** пол".
Поэтому назвал это поле именно так: это действие (т. е. поле `description`) в неопределённой совершенной форме, т. е. `_perfect_aspect`. 

Я предположил, что заказчику может понадобиться фраза которая будет выводиться в отчёте, например: "[Вася] **помыл** пол".  Тогда у нас появится ещё одно поле, с фразой с глаголом совершенного вида в форме прошедшего времени, `description_perfect_past` (тут вот как раз могут появиться ещё и спряжения). Или наоборот, захотят планировщик для детей и нужна будет фраза с глаголами в будущем времени, `description_perfect_future`, "Завтра Вася **вымоет** / **будет мыть** пол "). 

Вот такая логика :)

В принципе, можно подойти с другой стороны, не с грамматической формы, как я рассуждал, а с того, для чего нужно это поле и где оно будет использоваться, тогда — да, `description_for_message` лучше, чем придумал я (тогда, как развитие, будет `description_for_report`, `description_for_schedule` — норм 😉). 
Единственное "но", которое я вижу: а если в будущем это будет использоваться не только для сообщений, но и ещё для чего-то? Тогда это название будет путать.

ADesBiysk replied:
C Disclaimer полностью согласен. :) Когда описываешь задачу, то больше думаешь о технической части задачи. На остальное времени остаётся всё меньше. 
Сомневаюсь, что в будущем появится необходимость в других грамматических формах. Это сообщение отправляется пользователю, поэтому не должно раздражать или быть "корявым". В отчётах для внутреннего использования будет достаточно использовать поле description, которое будет вставляться в соответсвующую ячейку таблицы.

ivr42 replied:
ок, используем `description_for_message`

ADesBiysk commented on src/core/db/models.py:
```diff
@@ -105,8 +105,11 @@ class Task(Base):
+    description_perfect_aspect = Column(String(length=150), nullable=False)
     reports = relationship("Report", back_populates="task")
 
+    __table_args__ = (UniqueConstraint("description_perfect_aspect", name="uc_task_description_perfect_aspect"),)
```
Почему индекс создаётся отдельно, а не используется unique=True?

ivr42 replied:
Alembic, при автоматическом создании миграции, создаёт отдельно столбец и отдельно constraint к нему. Удаляет — также: сначала constraint, потом сам столбец. Я подумал, что раз они так делают — это более универсальное решение. Но в этом случае, нам надо знать имя constraint-а, собственно, поэтому я его и создаю отдельно. 
Но я не нашёл объяснения почему alembic так делает. Возможно для некоторых СУБД это важно? Я оставил в том виде, как это сделал alembic, и добавил имя для constraint-а. 
В принципе, для postrgresql это лишнее — при удалении столбца он сам удаляет constraint. 

Как лучше сделать: 1). Оставить как есть, или 2). убрать явное задание имени constraint-а + убрать явное удаление constraint-а в функции `downgrade` миграции?

ADesBiysk replied:
Если задать у нового поля unique=True, то результат будет аналогичный. Также будет создан индекс. 
Я бы использовал именно такой вариант поскольку:
1. Единообразие кода. В models.py индексы описываются, только если требуется обеспечить уникальность сочетаний нескольких полей одновременно.
2. Индекс будет создан автоматически. Не вижу смысла загружать код явным указанием того, что и так будет автоматически сделано "под капотом".

ivr42 replied:
хорошо, уберу лишнее

ADesBiysk commented on tasks.json:
```diff
@@ -1,31 +1,31 @@
-{"filename": "clean_the_teeth.jpg", "description": "Почисти зубы"},
-{"filename": "clean_up_the_closet.jpg", "description": "Убери в шкафу"},
-{"filename": "clean_your_shoes.jpg", "description": "Почисти обувь"},
+[{"filename": "change_the_bed_linen.jpg", "description": "Поменяй постельное бельё", "description_perfect_aspect": "Поменять постельное бельё"},
```
Текст нового поля должен начинаться со строчной буквы.
Сегодня твоим заданием будет поменять постельное бельё.

ivr42 replied:
ок

ADesBiysk reviewed (commented):
Нужно заменить поле, значение которого вставляется в сообщение.

ivr42 committed:
Merge branch 'develop' into feature/task_in_the_perfect_aspect

ivr42 committed:
Code-review fixes

  - `description_for_message` used now as a field name
  - automatic constraint name used in `Task` model and migrations
  - lower case used for values in `description_for_message` field

ADesBiysk reviewed (commented):
Нужно использовать новое поле при формировании сообщения пользователю. /src/bot/jobs.py, функция send_daily_task_job.

ivr42 committed:
Merge branch 'develop' into feature/task_in_the_perfect_aspect

ivr42 committed:
`description_for_message` is used now in `send_daily_task_job`

ADesBiysk reviewed (commented):
Добавился ещё один эндпойнт PATCH /tasks/{id}. Нужно в него тоже добавить новое поле.

Ivan-Skvortsov commented on tasks.json:
```diff
@@ -1,31 +1,31 @@
-{"filename": "hang_up_the_laundry.jpg", "description": "Развесь бельё"},
+{"filename": "comb_your_head.jpg", "description": "Причешись", "description_for_message": "причесаться"},
+{"filename": "hang_up_the_laundry.jpg", "description": "Развесь бельё", "description_for_message": "развесить бельё"},
 {"filename": "household_chores.jpg", "description": "Домашние дела"},
```
сделать домашние дела?

ivr42 replied:
Привет! description был кем-то (возможно заказчиком) сформулирован так.
Возможно они хотят использовать часть тасков как заголовки?
Если это некорректно, то надо получить от заказчика корректный справочник заданий.

Ivan-Skvortsov replied:
Все верно, `description` сформулирован заказчиком. А `description_for_message` мы формулируем сами, на основании `description`, разве нет?) 
Кстати, конкретно в этом случае, description_for_message вообще отсутствует.
Также, отсутствуют description_for_message в заданиях:
```
{"filename": "clean.jpg", "description": "Уборка"},
{"filename": "sleepwear.jpg", "description": "Одежда для сна"},
```

ivr42 replied:
поправил

ivr42 replied:
для тех заданий, где это поле отсутствует, а также для тех заданий, которые есть в базе, но их нет в файле `tasks.json`, значения в поле `description_for_message` формируются из значений в поле `description` путём преобразования последних к нижнему регистру.

ivr42 replied:
> А description_for_message мы формулируем сами, на основании description, разве нет?)

сейчас — да, т. к. вариантов нет
но в будущем, заказчик должен заполнять это поле сам

Ivan-Skvortsov commented on tasks.json:
```diff
@@ -1,31 +1,31 @@
+{"filename": "stroke_it.jpg", "description": "Погладь", "description_for_message": "погладить"},
+{"filename": "sweep.jpg", "description": "Подмети", "description_for_message": "подмести"},
+{"filename": "take_out_the_trash.jpg", "description": "Вынеси мусор", "description_for_message": "вынести мусор"},
+{"filename": "things_in_their_places.jpg", "description": "Вещи по местам", "description_for_message": "вещи по местам"},
```
"Сегодня твоим заданием будет вещи по местам" - не очень красиво звучит, на мой взгляд)
Может, "разложить вещи по местам"? Или "убрать вещи на свои места"?

ivr42 replied:
поправил

ivr42 replied:
для тех заданий, где это поле отсутствует, а также для тех заданий, которые есть в базе, но их нет в файле `tasks.json`, значения в поле `description_for_message` формируются из значений в поле `description` путём преобразования последних к нижнему регистру.

Ivan-Skvortsov commented on tasks.json:
```diff
@@ -1,31 +1,31 @@
+{"filename": "wash_the_floor.jpg", "description": "Помой пол", "description_for_message": "помыть пол"},
+{"filename": "wash_the_hands.jpg", "description": "Помой руки", "description_for_message": "помыть руки"},
+{"filename": "wash_the_head.jpg", "description": "Помой голову", "description_for_message": "помыть голову"},
+{"filename": "wash.jpg", "description": "Вымой", "description_for_message": "вымыть"},
```
Тут тоже какое-то странное задание) Но это больше к заказчикам вопрос, наверное

ivr42 replied:
поправил

ivr42 committed:
`description_for_message` field was added `to update_task` method

ivr42 commented:
поправил ендпоинт PATCH /tasks/{id} по комменту 
https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/232#pullrequestreview-1324965902

ivr42 committed:
docstring of `update_task` was changed

ivr42 committed:
`description_for_message` was added to 'strange' tasks
